### PR TITLE
Feature/uint support

### PR DESCRIPTION
**Pull request by Leon0402**
Hi Roland, 

I saw in my project that the lib currently fails for integer_unsigned values and I also saw that someone has made in the past attempts to implement it (#25), so I picked it up. Hopefully we can get this in this time :) 

It seems to work with my test I wrote. I see an issue though with it yet. The sqlite3 lib doesn't natively support uint values, all integers are signed (https://www.sqlite.org/datatype3.html), therefore there is actually no real unsigned support (it's rather so there, so using the lib will be easier ... no casts needed). I'm not quite sure though what happens if you actually throw in a value, which is out of range (i.e fits into unsigned, but doesn't fit into signed). 
Any thoughts on this?

And I kinda screwed up with git as I have changed from master and from the development branch (my last fetch Content stuff directly went into master rather than development). So I based it now on master as well (but got commits from development).

### CODE REVIEW

**Comment by rbock** [comment]
Hi Leon,

> I'm not quite sure though what happens if you actually throw in a value, which is out of range (i.e fits into unsigned, but doesn't fit into signed).
> Any thoughts on this?

Did you add a test to see what happens?

If sqlite3 does not natively support unsigned values, is it a good idea to let the library support it?

Here is my thinking:
One of the benefits of having a library that checks a lot of stuff at compile time is that it will tell you if you are doing something that will not work. For instance, sqlite3 does not support outer join. The library won't let you use outer join and it will tell you so at compile time.
So if sqlite3 does not support unsigned integer values, then the correct behavior for the library would be to not support unsigned integer as well (albeit with a better error message).

**Comment by Leon0402** [comment]
Sqlite3 seems to not really support types in any way: It says on the website, types are more of a recommendation. The "INTEGER" Type stands for a lot of common types in other databases including Unsigned Integer.

So technically sqlite3 stores everything as int64 values (or less). Storing unsigned values in no problem as long as it fits in it (i.e int32). So the interesting case is more, what happens if you store some uint64_t value, which is bigger than uint32_t. I actually tested it (see my last commit, if you want to try it out yourself) with the maximum uint64_t value. It will be converted into an int64_t and therefore will become to -1 (integer overflow). 
If you fetch it again from the database, the signed value will be converted back and therefore will be correct again. 
Not too bad, but also not quite optimal I think :-) 

I think only supporting uint32_t instead of uint64_t would be fine in general. I see the main problem with types, which are not fixed in size (i.e. size_t, int, long int ...), they will often have 32 bit, but could have 64 as well. So on some machines it will work without a cast and on some in won't. 

The difference I see between your example with "outer join" is that it's optional I believe (So you could replace it just by another join). But uint64_t not being supported by the lib, won't stop you from using them in your app, it's kind of essential. So perhaps there will just be a bunch of static_casts. Well it's explicit then, but does it really make the code any safer? 

I'm really not sure on this, but I tend to say that uint64_t should be supported and it's the job of the library user to make sure it's not too big (And the lib should support here by documenting it). Not supporting it doesn't really make the code any safer I believe.

File: src/bind_result.cpp [hunk 1]
```diff
@@ -86,6 +86,16 @@ namespace sqlpp
       *is_null = sqlite3_column_type(_handle->sqlite_statement, static_cast<int>(index)) == SQLITE_NULL;
     }
 
+    void bind_result_t::_bind_unsigned_integral_result(size_t index, uint64_t* value, bool* is_null)
+    {
+      if (_handle->debug)
+        std::cerr << "Sqlite3 debug: binding unsigned integral result " << *value << " at index: " << index
+                  << std::endl;
+
+      *value = sqlite3_column_int64(_handle->sqlite_statement, static_cast<int>(index));
```
Comment (rbock) [inline src/bind_result.cpp:95]: I think your point about static_cast is right, but it should be done here.
It would be explicit and safe, I believe.

File: src/prepared_statement.cpp [hunk 2]
```diff
@@ -151,6 +152,20 @@ namespace sqlpp
...
 
+    void prepared_statement_t::_bind_unsigned_integral_parameter(size_t index, const uint64_t* value, bool is_null)
+    {
+      if (_handle->debug)
+        std::cerr << "Sqlite3 debug: binding unsigned integral parameter " << *value << " at index: " << index
+                  << ", being " << (is_null ? "" : "not ") << "null" << std::endl;
+
+      int result;
+      if (not is_null)
+        result = sqlite3_bind_int64(_handle->sqlite_statement, static_cast<int>(index + 1), *value);
```
Comment (rbock) [inline src/prepared_statement.cpp:163]: Here we should also have an explicit static_cast, I think.

**Review by rbock — COMMENTED** [review]
Thanks for your comments. See inline.

File: src/bind_result.cpp (hunk 1 above)
Comment (Leon0402) [reply src/bind_result.cpp]: I agree, changed it :-)

File: tests/IntegralTest.cpp [hunk 3]
```diff
@@ -70,41 +60,38 @@ int main()
...
-      fp REAL
+  db.execute(R"(CREATE TABLE integral_sample (
+      signed_value INTEGER,
+      unsigned_value INTEGER
   ))");
 
-  // check for automatic conversion in serializer
-  uint64_t v = 17032080461028570721ULL;
-  auto v2 = static_cast<int64_t>(v);
+  // Supported range for unsigned value (as int64 is the maximum)
```
Comment (rbock) [inline tests/IntegralTest.cpp:68]: I might be wrong, but I thought the conversion is defined by the standard, in other words it should also work for `std::numeric_limits<uint64_t>::max()`?

If that is not guaranteed by the standard, can you point me to the relevant section?

Comment (Leon0402) [reply tests/IntegralTest.cpp:68]: I'm not entirely sure, what you mean here. But I suppose my comment above line 69 might have confused you. What I meant is if you put anything bigger than int64_t in the lib it will still work (see line 75) technically, but the data in the database will be wrong (at least if you don't know that it's an uint64_t).
So `std::numeric_limits<uint64_t>::max()` will be converted (in a definied way as far as I know) into int64_t and therefore there will be an integer overflow resulting in -1.
The test shows that it still works (i.e. you get in the end std::numeric_limits<uint64_t>::max() back), but the database will save it as -1 and therefore another lib (or a human), who don't know that this is an unsigned value will interpret it wrong. That's want I meant by unsupported.

**Comment by Leon0402** [comment]
I noticed I have to call value() on the result, otherwise the mac os build will complain. It seems like the result set cannot converted (implicitly) to my value there, while it can on other platforms. 
Any reason for this? Is it always saver to use the value() function on the result set?

**Comment by rbock** [comment]
Can you give a bit of context here? Like a piece of code? What is the warning you are seeing?

You can see the relevant code for the implicit conversion operator here: https://github.com/rbock/sqlpp11/blob/develop/include/sqlpp11/result_field_base.h#L116

The `value` is a couple of lines above.

**Comment by Leon0402** [comment]
Have a look here: 
https://travis-ci.com/github/rbock/sqlpp11-connector-sqlite3/builds/191114902

This was the output before my last commit. It failed on mac os, but with the .value() function added in my last commit it doesn't.

File: tests/IntegralTest.cpp (hunk 3 above)
Comment (rbock) [reply tests/IntegralTest.cpp:68]: I see what you mean. Maybe add another line or two as explanation?

**Comment by rbock** [comment]
IIUC, the compilation error was in the last line only, which is comparing a unsigned result field (which can be implicitly converted to uint64_t) and a size_t.
My assumption would be that on Apple, uint64_t and size_t are not considered the same. Then there are two implicit conversions necessary (from result field to uint64_t and then one of the numbers has to be converted to another type). I am not an expert on those rules, but I think that is where the compiler is giving up. I was also not able to reproduce the effect with a smaller example yet, so I am not convinced of this assumption.

You could try using `const auto` or `const uint64_t` instead of `std::size_t`. My guess is that that would compile.

File: tests/IntegralTest.cpp (hunk 3 above)
Comment (Leon0402) [reply tests/IntegralTest.cpp]: Is the documentation I just added clear enough? Do you miss anything?

**Comment by rbock** [comment]
Thanks for the update. Looks good to me.

Are you happy with it too? Or do want to investigate something before we merge?

**Comment by Leon0402** [comment]
I just updated it again and added more tests for different data types. And I already expected it: It fails for all data types, which aren't uint64_t or int64_t except on Linux with gcc (with clang it fails). So this is actually a general "bug". 

The easy workaround is just to use value() and do an explicit cast. But I still wonder if there is a way to fix this (in sqlpp11 directly I suppose). I don't quite understand when the operator is actually called you mentioned above (I actually never heard of _cpp storage_type and other things you used there).

**Comment by rbock** [comment]
Hi Leon,

Looking at the error message again, I am pretty sure I was able to simulate the problem (the templated code is hard to read):

```C++
#include <cstdint>

class X
{
    using _cpp_storage_type = uint64_t;
public:
  auto operator!=(_cpp_storage_type) const -> bool;
  operator _cpp_storage_type() const;
};

template <typename L, typename R>
auto require_equal(const L& l, const R& r) -> void
{
  if (l != r)
  {
  }
}

int main()
{
  require_equal(X{}, 17);
}
```

_cpp_storage_type is just a type alias. The ambiguity comes from having both, an implicit conversion operator and comparison operators.

To solve the ambiguity in your piece of code, we could

 - make the conversion operator explicit
 - remove the comparison operators
 - pass a uint64_t instead of unsigned int
 - use `.value()`

The first two would almost certainly break existing code.

WDYT?

**Comment by Leon0402** [comment]
Okay I think I got now what happended, let me elaborate on the options: 

- make the conversion operator explicit

I think this creates more issues than it solves as simple code like: 

Basic stuff such as `uint64_t i = X{}` doesn't work then and requires a static_cast, which seems bad to me. After all, the goal would be (in my opinion) that the user doesn't know that he is working with a wrapper class. 

- use .value()

This solution has the same drawback as the first one, it makes the code a little bit more unintuitive to work with. I also think this isn't really an alternative to solution 1, but rather be used together with it. In this case, we should probably either make the operator explicit (or even remove it), but then encourage users to always use value() ... it's still less intuitive, but at least easy to use and consistent. 
(I just ignored backwards compatibility for now) 

- pass a uint64_t instead of unsigned int

That seems difficult to me and perhaps even less intuitive than Option 1 / 2. 

- remove the comparison operators 

This sounds to me like the best solution, although I'm not sure if I understood the implications of this to be honest. The operator doesn't seem to do much except from converting and then calling the built in operators. That's pretty much the same that happens without the operator as well (just implicti) I think?
Maybe you could explain where this breaks backwards compatibility?

**Comment by rbock** [comment]
The following fails, for instance

```C++
#include <cstdint>
#include <string>
class X
{
    using _cpp_storage_type = std::string;
public:
  //auto operator!=(_cpp_storage_type) const -> bool;
  operator _cpp_storage_type() const;
};

template <typename L, typename R>
auto require_equal(const L& l, const R& r) -> void
{
  if (l != r)
  {
  }
}

int main()
{
  require_equal(X{}, "sample");
}
```

I have no idea how much code would be affected. But in those instance, you would be forced to use `.value()` again.

Also, regarding `.value()`: If you are following the AAA guideline (almost always auto), you will notice that

```C++
auto x = result.id; // does not compile, because implicit conversion is not happening
auto y = result.id.value(); // works
```

I have therefore seen quite a few of `.value()` calls in the wild.

The "real" solution would be to move to unwrapped values. I did that in the C++17 version, it would be possible to do so in sqlpp11, but would be a major undertaking.

**Comment by Leon0402** [comment]
Well then it seems, like we should always use (and encourage) value(). It seems like no matter what we do, value()  is the only solution, which will always work. 

Using no wrapper at all is even better obviously, but I don't think it's worth it to implement it for sqlpp11. The time would be better invested in making sqlpp17 production ready (which I know you haven't planned).

**Comment by rbock** [comment]
I would say "use .value() when implicit conversion is not working", but yes, using value() is sometimes required.

**Comment by Leon0402** [comment]
> I would say "use .value() when implicit conversion is not working", but yes, using value() is sometimes required.

The problem is though that it seems to be compiler  and platform dependent whether it is required or not. Even if you know about the scenarios where it could go wrong (which most people perhaps don't know), that's easy to do wrongly. And in general I believe, it's good to recommend things, which work always. Experiences users still have the option to not listen to advice :-) 

But anyway, this would be ready to merge from my side.

**Comment by rbock** [comment]
I see what you mean. Let me rephrase. 

I think there are two use-cases:

1. If you assign to a value or const reference of known type, you can rely on the implicit conversion.
2. In all other cases, you should use `.value()`.

Most of the code I have written with sqlpp11 transferred results into a struct or container of structs. Thus, most of the time, I relied on implicit conversion.

**Comment by rbock** [comment]
Merged.

Thanks for the discussion. Appreciate the attention to detail!

**Comment by Leon0402** [comment]
Ah I see that makes sense, thanks for the explanation!
